### PR TITLE
docs: add some verbiage for benchmark test rules

### PR DESCRIPTION
**Pull request by pgenera**
Commit Message: Add documentation about benchmark test bazel rules.
Additional Description: I'm trying to communicate that envoy_cc_benchmark_binary is for humans measuring performance, and envoy_benchmark_test is for machines running CI builds.
Risk Level: None. Documentation only.
Testing: automated spell check.
Release Notes: N/A

### CODE REVIEW

**Comment by pgenera** [comment]
Hey @antoniovicente, do you think this will help the next time someone is confused by the difference between a benchmark_binary and benchmark_test rule?

File: test/README.md [hunk 1]
```diff
@@ -119,3 +119,19 @@ test infrastructure that wants to be agnostic to which `TimeSystem` is used in a
...
+`envoy_benchmark_test`, to execute them locally and in CI environments
+respectively. `envoy_benchmark_test` rules call the benchmark binary from a
+[script](https://github.com/envoyproxy/envoy/blob/master/bazel/test_for_benchmark_wrapper.sh)
+which runs the test quickly instead of rigourously. In order to collect
+meaningful bechmarks, `bazel run -c opt` the benchmark binary target on a
+quiescent machine.
+
+If you would like to detect when your benchmark test is running under the
+wrapper, call
+[`Envoy::benchmark::skipExpensiveBechmarks()`](https://github.com/envoyproxy/envoy/blob/master/test/benchmark/main.h).
```
Comment (antoniovicente) [inline test/README.md:137]: Could we add some logging when this flag is set to improve discoverability of the benchmark vs test issue?  The comment may want to say something about benchmark being run in test mode and point to this documentation on how to run the standalone benchmark.

https://github.com/envoyproxy/envoy/pull/11442#discussion_r456681014

File: test/README.md [hunk 2]
```diff
@@ -119,3 +119,19 @@ test infrastructure that wants to be agnostic to which `TimeSystem` is used in a
...
+
+
+## Benchmark tests
+
+Envoy uses [Google Benchmark](https://github.com/google/benchmark/) for
+microbenchmarks. There are custom bazel rules, `envoy_cc_benchmark_binary` and
+`envoy_benchmark_test`, to execute them locally and in CI environments
+respectively. `envoy_benchmark_test` rules call the benchmark binary from a
+[script](https://github.com/envoyproxy/envoy/blob/master/bazel/test_for_benchmark_wrapper.sh)
+which runs the test quickly instead of rigourously. In order to collect
```
Comment (antoniovicente) [inline test/README.md:131]: Possible alternate wording suggestion:
which runs the benchmark with a minimal number of iterations and skipping expensive benchmarks to quickly verify that the binary is able to run to completion.

File: bazel/envoy_test.bzl [hunk 3]
```diff
@@ -252,7 +253,9 @@ def envoy_cc_benchmark_binary(
         **kargs
     )
 
-# Tests to validate that Envoy benchmarks run successfully should be specified with this function.
+# Tests to validate that Envoy benchmarks run successfully should be specified
+# with this function. Not for actual performance measurements: iteratons will be
+# skipped in the interest of execution time.
```
Comment (antoniovicente) [inline bazel/envoy_test.bzl:258]: and expensive benchmarks will be skipped

File: test/README.md (hunk 1 above)
Comment (pgenera) [reply test/README.md:138]: Done; although I couldn't use any of the fun ENVOY_LOG macros outside of a class. It logs every call, which constitutes "big nice warning" to me ;)

File: test/benchmark/main.cc [hunk 4]
```diff
@@ -37,4 +39,11 @@ int main(int argc, char** argv) {
   benchmark::RunSpecifiedBenchmarks();
 }
 
-bool Envoy::benchmark::skipExpensiveBenchmarks() { return skip_expensive_benchmarks; }
+bool Envoy::benchmark::skipExpensiveBenchmarks() {
+  if (skip_expensive_benchmarks) {
+    auto logger = Logger::Registry::getLog(Logger::Id::testing);
+    logger.warn("Expensive benchmarks are being skipped; see test/README.md for more information");
```
Comment (jmarantz) [inline test/benchmark/main.cc:45]: ENVOY_LOG_MISC

File: test/benchmark/main.cc [hunk 5]
```diff
@@ -37,4 +39,11 @@ int main(int argc, char** argv) {
   benchmark::RunSpecifiedBenchmarks();
 }
 
-bool Envoy::benchmark::skipExpensiveBenchmarks() { return skip_expensive_benchmarks; }
+bool Envoy::benchmark::skipExpensiveBenchmarks() {
+  if (skip_expensive_benchmarks) {
+    ENVOY_LOG_MISC(
+        warn, "Expensive benchmarks are being skipped; see test/README.md for more information");
```
Comment (antoniovicente) [inline test/benchmark/main.cc:45]: I would just log in main before the call to benchmark::RunSpecifiedBenchmarks()

Also, we may want to LOG at critical level so for visibility.

Comment (pgenera) [reply test/benchmark/main.cc]: Done.

**Pull request merged**